sam/bai.go

```diff
@@ -0,0 +1,295 @@
+// random access of reads given a query genomic range. The Bai
+// struct should not be accessed directly, but used through
+// query functions present within the sam package of gonomics.
+type Bai struct {
+	refs []reference
+
+	hasNoCoordReads bool   // true if optional field below exists
+	noCoordReads    uint64 // number of unplaced unmapped reads
+}
+
+// reference sequence being binned.
+type reference struct {
+	sliceIdx    uint32 // index in Bai.refs
+	head        bin    // a dummy bin at the top of tree to organize otherwise unlinked bins.
+	bins        []bin
+	intervalOff []uint64
+	noCoord     noCoordBin
+}
+
+// bin containing all reads that are contained within a variable genomic size.
+// bins are organized into a tree structure for efficient queries by genomic coordinates.
+type bin struct {
+	sliceIdx uint32 // index in reference.bins
+	id       uint32 // distinct bin id (only unique within reference). see const for id -> size.
+	refStart uint32 // genomic start coordinate of bin (0-base, closed)
+	refEnd   uint32 // genomic end coordinate of bin (0-base, open)
+	chunks   []chunk
+	parent   *bin
+	children []*bin
+}
+
+// chunk represents the start and end virtual offset of a contiguous
+// block of bam records that fall within a given bin. All records within
+// a bin are not necessarily contiguous, therefore a bin may have multiple chunks.
+type chunk struct {
+	start uint64 // virtual offset to start of chunk
+	end   uint64 // virtual offset to end of chunk
+}
+
+// noCoordBin is a special 'dummy' bin with id == 37450 that stores
+// metadata about the number of mapped/unmapped reads for a given reference
+// as well as the start and end virtual offsets of the unmapped reads in
+// the bam file.
+type noCoordBin struct {
+	start       uint64 // virtual offset to start of chunk
+	end         uint64 // virtual offset to end of chunk
+	numMapped   uint64
+	numUnmapped uint64
+}
+
+// ReadBai reads an entire bai file into a bytes buffer and parses
+// the buffer into a Bai struct which can be used for efficient
+// random access of reads in a bam file.
+func ReadBai(filename string) Bai {
+	file := fileio.EasyOpen(filename)
+	r := new(bytes.Buffer)
+	_, err := r.ReadFrom(file) // read entire file to bytes buffer
+	exception.PanicOnErr(err)
+	err = file.Close() // close read file
+	exception.PanicOnErr(err)
+
+	if string(r.Next(4)) != magicBai {
+		log.Fatalf("malformed bai header in '%s'", filename)
+	}
+
+	bai := parseBai(r)
+
+	if r.Len() == 8 { // check for presence of optional field
+		bai.hasNoCoordReads = true
+		bai.noCoordReads = le.Uint64(r.Next(8))
+	}
+
+	if r.Len() != 0 {
+		log.Fatalf("ERROR: extra '%d' bytes found in bai file.\nfile may be malformed: '%s'\n", r.Len(), filename)
+	}
+
+	for i := range bai.refs {
+		bai.refs[i] = annotateBinRanges(bai.refs[i])
+		bai.refs[i] = assembleTree(bai.refs[i])
+	}
+	return bai
+}
+
+// alias for ease of use and readability
```

Comment:
This might make it easier for people who are reading the entire file top-to-bottom, but maybe it makes it harder for people who get an error and jump to a specific function/line to see what is going on (e.g. parseBin).  Then they need to either scroll around or use an smart IDE feature to figure out what "le" refers to.

Reply:
It makes things much nicer to read imo, but I do agree with you. How do you feel about this compromise? I've moved the alias to the top of the file, and added a note to the first instance of `le` in each function so others would only need to look to the first use of `le` like they would for any variable declaration.